### PR TITLE
Add date of birth to add mentor

### PR DESCRIPTION
**Pull request by Kizr**
## Context

New requirements mean that we need to request the date of birth from a mentor before searching for them in the service.

## Changes proposed in this pull request

- [x] Updates the add a mentor flow to include date of birth
- [x] Updates the support add a mentor flow to include date of birth

## Guidance to review

Check out the code and try adding a mentor as Anne and Colin.

## Link to Trello card

["Add a mentor" – add DOB to the flow](https://trello.com/c/MqTIGRYi)

## Screenshots

image
image
image

### CODE REVIEW

**Review by CatalinVoineag — APPROVED** [review]
Looks good to me 👍🏻

**Review by ollietreend — APPROVED** [review]
I'm approving this because it's effectively a straight copy-and-paste of what's been implemented on the Claims side (in #542). So let's go with it for now, in order to get this change over the line in a timely manner.

However, I think this is something we should come back to later and tidy up. In particular I don't think the decision to duplicate code between both services is sensible given they behave the same. In general, one of our principles for the shared codebase is that implementations should be DRY where possible, and then we namespace into individual services at the point of difference.

@Nitemaeric I'm going to write up a tech-debt card for this. Let's work out which team has capacity to pick it up and when.

File: app/services/mentor_builder.rb [hunk 1]
```diff
@@ -1,27 +1,30 @@
...
     @first_name = first_name
     @last_name = last_name
+    @date_of_birth = date_of_birth
   end
 
   def call
-    mentor = Mentor.find_or_initialize_by(trn:)
-    return mentor if mentor.persisted? # We already have the data (first_name, last_name), so don't need to fetch or assign
-    return mentor if mentor.invalid_attributes?(:trn) # If the trn doesn't meet basic validation, we can't build any further
+    mentor = Mentor.find_or_initialize_by(trn:).becomes(Placements::Mentor)
```
[the remark this answers is not in the document]
Comment (Nitemaeric) [reply app/services/mentor_builder.rb:12]: We had to diverge as we added in new functionality. We didn't know if Placements would have the same requirements.

To DRY this up, we could add a new method, `find_or_initialize_mentor` that can be overridden in each service's subclass implementation.

**Comment by Nitemaeric** [comment]
> In general, one of our principles for the shared codebase is that implementations should be DRY where possible, and then we namespace into individual services at the point of difference.

Agreed. That being said, it's sometimes difficult to anticipate shared behaviour until both services require the shared behaviour.

As you've picked up here, once we have shared behaviour, we can extract out the common behaviours into a common class/partial/component.

**Comment by ollietreend** [comment]
Tech debt card: https://trello.com/c/oPcsVtuK/146-refactor-the-mentorbuilder-to-dry-up-duplicate-implementations

**Pull request merged**